[Init] Initial commit for Synaptics SL1680 NPU

GaryHuang-ASUS opened the pull request:
* add a rough detector which is testing with mobilenet synap model.

## Proposed change
<!--
  Thank you!

  If you're introducing a new feature or significantly refactoring existing functionality,
  we encourage you to start a discussion first. This helps ensure your idea aligns with
  Frigate's development goals.

  Describe what this pull request does and how it will benefit users of Frigate.
  Please describe in detail any considerations, breaking changes, etc. that are
  made in this pull request.
-->

Hello, 
We want to add a new type of detector which is [Synaptics SL1680 NPU](https://www.synaptics.com/assets/product-brief/astra-sl1680-embedded-iot-processor) .

Currently, we're testing it with Frigate NVR version 0.16.0 and mobilenet tflite model
This initial commit serves as the beginning of the upcoming work. 

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code
- [ ] Documentation Update

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:

## Checklist

<!--
  Put an `x` in the boxes that apply.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] UI changes including text have used i18n keys and have been added to the `en` locale.
- [ ] The code has been formatted using Ruff (`ruff format frigate`)

GaryHuang-ASUS committed:
[Init] Initial commit for Synaptics SL1680 NPU

* add a rough detector which is testing with yolov8 tflite model.

hawkeye217 commented:
Thanks for this PR. 

A few things:
- You indicated that you are just testing with yolov8, so this is relevant: https://github.com/blakeblackshear/frigate/pull/10717
- Once you have your code finalized and tested, you'll need to add documentation to the community supported boards section of the Object Detectors documentation: https://docs.frigate.video/configuration/object_detectors/#setup-1 

Go ahead and use this PR to test, finalize, and document your community supported detector. We'll review when you are ready.

GaryHuang-ASUS committed:
Merge remote-tracking branch 'upstream/dev' into Synaptics-SL1680

NickM-27 commented:
@GaryHuang-ASUS I believe the recommendation would be to start with a "community supported build" (see rk RockChip or RPi for examples) since this is targeting specific hardware. But the general approach would be the same. 

Regarding the error you got, you likely need to make sure you have docker buildx installed

NickM-27 commented:
I would assume that the dependencies may already be installed on the hardware that this runs on, so user_instsllation may not be necessary

hawkeye217 commented:
> I would assume that the dependencies may already be installed on the hardware that this runs on, so user_instsllation may not be necessary?

Right. 

Since this detector will be for very specific hardware, a community supported build like `rk` or `rpi` makes the most sense.

GaryHuang-ASUS committed:
[Feat] Add dependencies installation in docker build

- Add runtime library and wheels installation in main/Dockerfile
- Add model.synap(default model, transfer from mobilenet_224full80) in docker/synap1680

GaryHuang-ASUS committed:
[Update] Remove dependencies installation from main Dockerfile

- remove deps installation from Dockerfile
- add dependencies installation and split wheels, deps stage in synap1680 Dockerfile

GaryHuang-ASUS commented on docker/synap1680/install_deps.sh:
```diff
@@ -0,0 +1,15 @@
+
+apt-get -qq update
+
+if [[ "${TARGETARCH}" == "amd64" ]]; then
```
Oops! After checking, it seems only supports arm64. I'll update it later.

GaryHuang-ASUS replied:
Hi @NickM-27 ,

I've remove unused script(install_deps.sh and Dockerfile toolchain).

The runtime libraries installation issue is still checking with the vendor, so there might still be some changes here.

I also update the description about Synaptics SL1680 in those setup/configuration documents.

NickM-27 committed:
Refactor synap detector to more closely match other implementations

NickM-27 commented on frigate/detectors/plugins/synap1680.py:
```diff
@@ -0,0 +1,103 @@
+        self.model_shape = self.network_input_details.shape
+        self.model_dtype = self.network_input_details.data_type
+
+    def detect_raw(self, tensor_input):
```
I have gone ahead and made the changes as I now have the unit for testing

GaryHuang-ASUS replied:
> ore specific suggestio

Okay, I'll survey the part of ModelType and update it. 
It also needs to be mentioned in Frigate NVR official document, right? 

For the part of post-processing, we only have the pre-converted model (mobilenet 80 convert from .tflite to .synap, model shape as nhwc) and test it currently. The model type which is supported is mentioned in official document https://synaptics-synap.github.io/doc/v/latest/docs/manual/working_with_models.html#model-conversion, but we don't have those model for testing yet.

Thanks.

GaryHuang-ASUS committed:
[Update] Add model path configuration check

GaryHuang-ASUS committed:
[Update] update ModelType to ssd

GaryHuang-ASUS committed:
[Update] Remove unuse script

- install_deps.sh has already been executing in deps download stage
- Dockerfile.toolchain is for testing to extract runtime libraries from Synaptics toolchain

GaryHuang-ASUS committed:
[Update] update Synaptics SL1680 setup description

GaryHuang-ASUS committed:
[Update] remove install_synap1680

- The deps download and installation is existed in synap1680

GaryHuang-ASUS committed:
[Fix] update document content

GaryHuang-ASUS committed:
[Update] Update detector from synap1680 to synaptics

This update is in order to make the synaptics SL-series NPU detector more general.

- Fix detector `os` module not import bug
- Update detector type `synap1680` to `synaptics`
- Update document description `SL1680` to `Synaptics` only
- Update docker build content `synap1680` to `synaptics`

GaryHuang-ASUS committed:
[Fix] Update configuration document

GaryHuang-ASUS commented:
> You mean something similar to what is done for building Hailo drivers in a dedicated repo? I don't have a problem doing that as long as I can confirm the licenses permit it.

Hi @blakeblackshear ,

    Synaptics has already replied to us saying that they have changed the library-related licenses to Apache 2.0. Therefore, it should now be possible to host them for download in your (frigate nvr) GitHub release, similar to how it is done for hailort. Once Frigate team done this(fork and put necessary files into repo release), I will update the current dependency download path in docker/synaptics/Dockerfile.

The runtime libraries can be retrieve from their prebuild toolchain's sysroot or compile it and get it from their sdk.
[prebuild toolchain](https://github.com/synaptics-astra/sdk/releases/tag/v1.5.0)
[compile from their sdk](https://synaptics-astra.github.io/doc/v/1.5.0/yocto.html#how-to-build-an-image)

Following is the information of their revision for license and response content for us: 
[license update information](https://github.com/synaptics-synap/runtime/pull/5)

```
Hi GaryHuang,

libebg_utils.so and libsynapnb.so are now Apache 2.0. We relicensed these files a while ago, but did not update the headers or recipes. Please consider these files Apache 2.0. We will update the source code in a future release. 

Since these libraries are Apache 2.0 there is no restriction on uploading rebuilt libraries on GitHub.
```

Thanks.

GaryHuang-ASUS commented on docs/docs/configuration/object_detectors.md:
```diff
@@ -838,6 +842,35 @@ model:
+
+#### SSD
+
+Use this configuration for ssd models. Here's a default pre-converted ssd model under the root folder.
```
How would you prefer this to be described in the documentation? e.g Remind the user to configure it? or something else?

I will make the revision together with the suggestions above.

Thanks.

GaryHuang-ASUS committed:
Update docs/docs/configuration/object_detectors.md

Co-authored-by: Nicolas Mowen <nickmowen213@gmail.com>

GaryHuang-ASUS committed:
[Update] Update document content and detector default layout

- Update object_detectors document
- Update detector's default layout
- Update default model name

GaryHuang-ASUS committed:
[Update] Update object detector document content

GaryHuang-ASUS committed:
[Fix] Fix InputTensorEnum not defined error

- import InputTensorEnum from detector_config

NickM-27 reviewed (changes requested):
This looks good, just a few things needed:
1. Fix the failing python checks
2. Need to add the build to .github/workflows/ci.yml

GaryHuang-ASUS committed:
[Update] Update detector script coding format

GaryHuang-ASUS committed:
[Update] Update synaptics detector coding format

GaryHuang-ASUS commented:
> This looks good, just a few things needed:
> 
> 1. Fix the failing python checks
> 2. Need to add the build to .github/workflows/ci.yml

Okay. I've already update the python checks and I'm going to add the build to ci.yml. It will be update later.

Thanks.

GaryHuang-ASUS committed:
[Update] Add synaptics ci workflow

GaryHuang-ASUS committed:
[Update] update synaptics runtime libs download path

- Fork Synaptics astra sdk repo and put the runtime lib package on it
- Frigate team can update this download path later

hawkeye217 commented:
Looks good. Thanks for your contribution!

GaryHuang-ASUS commented:
> Looks good. Thanks for your contribution!

Thanks for your kind and support!
